keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,532 @@
+   },
+   "status": "ready"
+}
+```
+
+And
+
+```go
+type Request struct {
+  Method  string
+  Timeout int
+  secret  string
+  Con     *Connection
+}
+req := Request{Method: "GET", Timeout: 30, secret: "pony"}
+klog.InfoS("Request finished", "request", Request)
+```
+
+That would result in log (pretty printed)
+```json
+{
+   "ts": 1580306777.04728,
+   "v": 4,
+   "msg": "Request finished",
+   "request":{
+      "Method": "GET",
+      "Timeout": 30
+   }
+}
+```
+
+
+### Logging configuration
+
+To allow selection of the logging output format we would like to introduce a new logging configuration shared by all kubernetes components.
+
+`LoggingConfig` structure should be implemented as part of `k8s.io/component-base` options and include a common set of flags for logger initialization.
+
+`--logging-format` flag should allow you to pick between logging output formats. Setting this flag will select a particular logger implementation.
+
+Proposed flag `--logging-format` values:
+* `text` for text-based logging format (default)
+* `json` for new JSON format
+
+Additionally we propose to expose to allow for customization of logging by providing custom implementation.
+Custom logger format can be introduced by implementing [https://github.com/go-logr/logr](https://github.com/go-logr/logr) interface.
+Registering custom implementation within `LoggingConfig` will allow to pick it by via `--logging-format` flag.
+Klog interface was selected as it is already supported by klog.v2 with `SetLogger` method.
+
+### Migration / Graduation Criteria
+
+#### Alpha
+
+Preparation and migration of selected logs to new logging methods:
+* Support for the new InfoS and ErrorS method will be implemented in klog and the new version of klog will be released.
+* Kubernetes will be upgraded to use the new klog version.
+* Helper methods for producing consistent Kubernetes resource identifiers to be used in logs will be implemented.
+* Most important klog calls will be identified and will be manually migrated from using format based to structured methods.
+
+#### Beta
+
+Adding support for JSON output format and guarding against regression:
+* klog version used by Kubernetes and all of its dependencies will be upgraded to v2.
+* Implementation of logr which will produce logs in JSON format will be created.
+* Flag which will inject new logr implementation into klog if --logging-format is set to json will be added to k8s.io/component-base JSON format
+* Static analysis protects against adding new string formatting logs
+* [Kubernetes Logging convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md) document is updated
+
+#### GA
+
+* All feedback for text and json logs output formats is addressed and both output formats become an API.
+
+### Performance
+
+Proposed changes could impact k8s performance in two ways:
+* By introducing degrading in klog implementation
+* By largely increasing log volume
```

Comment:
I know changing what is logged is outside of scope, but if we think ability to address resource across time and space is inevitable, I would prefer that we include uid in the first pass so we know the volume of logs produced can uniquely identify objects.

Reply:
Still in many cases the only information available in function scope is name. Logging UID would require larger code changes to have access to it, which are out of the scope.